### PR TITLE
refactor: change state structure for chunks

### PR DESCRIPTION
**Pull request by crepererum**
This is the first step towards #1513. However it leaves all consumers
bascially unchanged and also does NOT touch state transitions. These
changes will follow in upcoming PRs.

### CODE REVIEW

File: server/src/db/catalog/chunk.rs [hunk 1]
```diff
@@ -37,40 +37,108 @@ impl ChunkLifecycleAction {
...
 
-    /// Chunk is closed for new writes
-    Closed(Arc<MBChunkSnapshot>),
+/// Chunk in _persisted_ stage.
+#[derive(Debug)]
+pub struct ChunkStagePersisted {
+    /// Parquet chunk that lives immutable within the object store.
+    pub parquet: Arc<ParquetChunk>,
+    
+    /// In-memory version of the parquet data.
```
Comment (alamb) [inline server/src/db/catalog/chunk.rs:77]: I like this formulation of an optional readbuffer cache w/ ParquetChunk as the core representation -- perhaps eventually we will add a local file cache of the parquet file here as well

File: server/src/db/catalog/chunk.rs [hunk 2]
```diff
@@ -378,35 +457,38 @@ impl Chunk {
...
-        match &mut self.state {
-            ChunkState::Open(chunk) => Ok(chunk),
-            state => unexpected_state!(self, "mutable buffer reference", "Open or Closed", state),
+        match &mut self.stage {
+            ChunkStage::Open(stage) => Ok(&mut stage.mb_chunk),
+            stage => unexpected_state!(self, "mutable buffer reference", "Open or Closed", stage),
         }
     }
 
     /// Set the chunk to the Closed state
```
Comment (alamb) [inline server/src/db/catalog/chunk.rs:466]: maybe this should be renamed to `set_frozen` as there is no longer any notion of a `closed` chunk

File: server/src/db/catalog/chunk.rs [hunk 3]
```diff
@@ -472,65 +578,118 @@ impl Chunk {
         &mut self,
```
Comment (alamb) [inline server/src/db/catalog/chunk.rs:578]: the docstring on this method as gotten stale (there is no more MovingToObjectStore state)

File: server/src/db/catalog/chunk.rs [hunk 4]
```diff
@@ -472,65 +578,118 @@ impl Chunk {
...
+                    self.metrics.immutable_chunk_size.observe_with_labels(
+                        stage.parquet.size() as f64,
+                        &[KeyValue::new("state", "os")],
+                    );
+
+                    let rub_chunk = Arc::clone(read_buffer);
+                    stage.read_buffer = None;
+                    Ok(rub_chunk)
+                } else {
+                    // TODO: do we really need to error here or should unloading an unloaded chunk be a no-op?
```
Comment (alamb) [inline server/src/db/catalog/chunk.rs:679]: I agree this doesn't really seem to warrant an error. Maybe we could rename the method to `clear_memory_cache` or something to be clearer

**Review by alamb — APPROVED** [review]
I really like this structure 👍 @crepererum 

This PR will likely conflict fairly hard with ~https://github.com/influxdata/influxdb_iox/pull/1585~ https://github.com/influxdata/influxdb_iox/pull/1575 but I am happy to have this PR merged first and handle the merge conflicts

File: server/src/db/catalog/chunk.rs (hunk 1 above)
Comment (crepererum) [reply server/src/db/catalog/chunk.rs:77]: That was exactly the vision I had in mind when designing this struct :+1:

File: server/src/db/catalog/chunk.rs (hunk 2 above)
Comment (crepererum) [reply server/src/db/catalog/chunk.rs:466]: Let me do that in the next PR(s). I wanted to keep the impact radius (and the potential merge conflicts) as small as possible and therefore left all methods as they work now.

File: server/src/db/catalog/chunk.rs (hunk 4 above)
Comment (crepererum) [reply server/src/db/catalog/chunk.rs:679]: As for the other, similar case: let me keep the method names and functionality for this PR so I can adjust them piece-by-piece in upcoming PRs.

File: server/src/db/catalog/chunk.rs (hunk 3 above)
Comment (crepererum) [reply server/src/db/catalog/chunk.rs:578]: fixed

**Comment by crepererum** [comment]
> This PR will likely conflict fairly hard with #1585 but I am happy to have this PR merged first and handle the merge conflicts

@alamb thanks a lot!

File: server/src/db/catalog/chunk.rs (hunk 4 above)
Comment (alamb) [reply server/src/db/catalog/chunk.rs:679]: Sounds like a good plan

**Pull request merged**